sdk/core/azure-core/inc/azure/core/internal/hkeyholder.hpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Internal utility functions for strings.
+ *
+ */
+#pragma once
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+
+#include <windows.h>
+
+namespace Azure { namespace Core { namespace Internal {
+
+  class HKeyHolder {
+  private:
+    HKEY m_value;
+
+  public:
+    HKeyHolder() : m_value(NULL) {}
+
+    ~HKeyHolder()
+    {
+      if (m_value != NULL)
+      {
+        ::RegCloseKey(m_value);
+      }
+    }
+
+    void operator=(HKEY p)
+    {
+      if (p != NULL)
+      {
+        m_value = p;
+      }
+    }
+
+    operator HKEY() { return m_value; }
+
+    operator HKEY*() { return &m_value; }
+
+    HKEY* operator&() { return &m_value; }
```

Comment:
Do we need all these? I don't see where they are being called. Can we one method called `GetHKey()` or `GetValue() that returns the `HKEY`?

Reply:
These are used by the methods that access the HKEY.  They allow you to interact with the key as if its an HKEY.
For example this usage
```
    Azure::Core::Internal::HKEYHolder regKey;
    if (RegOpenKeyExA(
            HKEY_LOCAL_MACHINE,
            "SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion",
            0,
            KEY_READ,
            &regKey)
```
The last parameter is an HKEY.  The methods above allow us to pass an HKEYHolder ref to that function because the operators allow the cast to HKEY to do the right thing.